src/frontend/src/optimizer/rule/apply_agg.rs

```diff
@@ -48,9 +49,25 @@ impl Rule for ApplyAggRule {
             agg_call.filter = agg_call.filter.clone().rewrite_expr(&mut shift_index);
         });
 
-        let new_apply = apply.clone_with_left_right(apply.left(), input);
-        let new_agg = LogicalAgg::new(agg_calls, group_key, new_apply.into());
-        Some(new_agg.into())
+        let new_apply = LogicalApply::create(
+            left,
+            input,
+            join_type,
+            Condition {
+                conjunctions: vec![],
+            },
+            correlated_id,
+            correlated_indices,
+        );
+        let new_agg: PlanRef = LogicalAgg::new(agg_calls, group_key, new_apply).into();
+
+        // left apply's on condition for predicate push to deal with
+        if !on.conjunctions.is_empty() {
+            let filter = LogicalFilter::create(new_agg, on);
+            Some(filter)
+        } else {
+            Some(new_agg)
+        }
```

Comment:
`LogicalFilter::create(new_agg, on); `is enough because `create` won't produce `LogicalFilter` if `on` is always true.

BTW, can `on` contain `correlated_input_ref`?

Reply:
The `on` condition of Apply will never contain correlated_input_ref. Here is the simple condition like join (by now only happen after ApplyFilterRule). Push the on condition down the aggregation is a nontrivial operation.